server/src/main/java/org/elasticsearch/action/search/SearchDfsQueryThenFetchAsyncAction.java

```diff
@@ -98,4 +102,19 @@ protected SearchPhase getNextPhase(final SearchPhaseResults<DfsSearchResult> res
             context
         );
     }
+
+    private ShardSearchRequest rewriteShardSearchRequest(ShardSearchRequest request) {
```

Comment:
I actually just spoke about this same logic with @javanna earlier today. I admit this is not easy to follow, and I'm certainly open to suggestions for improvement. The logic is here to support accurately gathering terms statistics. We could instead go through each query potentially as part of the dfs phase on the shard. I'm not sure that's necessarily simpler. I think we would probably want to convert this to a BooleanQuery either way to try to avoid gathering multiple term stats on the shard. No matter what we decide, we do have to address this prior to having the knn queries ready to integrate.

Reply:
I'd try to have the linear combination in a single place, which is easier to follow and that is the query used for aggs, rather than requiring the linear combination like here for dfs, which does not include knn. I am not entirely sure about the implications, curious to hear your findings Jack.

Reply:
With the removal of default linear combination I was able to move this into a single method as part of `AbstractSearchAsyncAction`. The rewrite behavior is now identical for dfs-query-then-fetch and query-then-fetch.

Reply:
I apologize as I didn't explain this well. I meant for accurate term stats you still need to process multiple queries in `DFSPhase` in some way prior to processing completed knn searches in the `DFSQueryPhase`.

You are correct that we could iterate through multiple queries and combine the term stats for the same results. But that is more complex than having a single boolean query created on the coordinator prior to hitting the `DFSPhase` on the shards.

Reply:
I figured out a way to move this to the shard in a single location in `parseSource`. I described it in one my high-level comments below. Hopefully, this reduces some of the complexities here.